fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/PruneComplexTypeUtil.java

```diff
@@ -0,0 +1,253 @@
+
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+public class PruneComplexTypeUtil {
+    // For example, we have a column col: MAP<INT, STRUCT<a INT, b STRUCT<c INT, d INT>>>
+    // And we have a sql: SELECT map_values(col).b.d, map_keys(col) from TABLE;
+    // Column col has a specific ColumnRefOperator
+    // map_values(col).b.d has three access paths:
+    // --- Access path 1: ComplexTypeAccessPath(MAP_VALUES)
+    // --- Access path 2: ComplexTypeAccessPath(STRUCT_SUBFIELD, "b")
+    // --- Access path 3: ComplexTypeAccessPath(STRUCT_SUBFIELD, "d")
+    // So it's access paths is: [Access path 1, Access path 2, Access path 3]
+    // map_keys(col) has one access paths:
+    // --- Access path 4: ComplexTypeAccessPath(MAP_KEYS)
+    // So it's access paths is: [Access path 4]
+    // Summary: ColumnRefOperator(col) has two access paths, we use access group to represent it.
+    // ColumnRefOperator(col):
+    // --- Access group: [[Access path 1, Access path 2, Access path 3], [Access path 4]]
+    protected static class Context {
+        // The same ColumnRefOperator may have multiple access paths for complex type
+        private final Map<ColumnRefOperator, List<ImmutableList<ComplexTypeAccessPath>>> accessGroups;
+
+        private boolean enablePruneComplexTypes;
+
+        public Context() {
+            this.accessGroups = new HashMap<>();
+            this.enablePruneComplexTypes = true;
+        }
+
+        public void setEnablePruneComplexTypes(boolean enablePruneComplexTypes) {
+            this.enablePruneComplexTypes = enablePruneComplexTypes;
+        }
+
+        public boolean getEnablePruneComplexTypes() {
+            return this.enablePruneComplexTypes;
+        }
+
+        public void addAccessPaths(ColumnRefOperator columnRefOperator,
+                                   ImmutableList<ComplexTypeAccessPath> accessPaths) {
+            accessGroups.putIfAbsent(columnRefOperator, new ArrayList<>());
+            accessGroups.get(columnRefOperator).add(accessPaths);
+        }
+
+        public void addAccessPaths(ColumnRefOperator columnRefOperator,
+                                   ImmutableList<ComplexTypeAccessPath> curAccessPaths,
+                                   List<ImmutableList<ComplexTypeAccessPath>> visitedAccessGroup) {
+            int size = visitedAccessGroup.size();
+            // We should we below loop to avoid ConcurrentModificationException
+            for (int i = 0; i < size; i++) {
+                addAccessPaths(columnRefOperator, concatAccessPaths(visitedAccessGroup.get(i), curAccessPaths));
+            }
+        }
+
+        public void add(ColumnRefOperator outputColumnRefOperator, ScalarOperator scalarOperator) {
+            List<ImmutableList<ComplexTypeAccessPath>> visitedAccessGroup = null;
+            if (outputColumnRefOperator != null) {
```

Comment:
when outputColumnRefOperator could be null?

Reply:
Like `Predicate` in `Operator`, it has no output columnRefOperator, it only has a `ScalarOperator` (Base class of `Predicate`).